pkg/fetch/openeuler/csaf/csaf.go

```diff
@@ -0,0 +1,216 @@
+	opts.wait = int(r)
+}
+
+func WithWait(wait int) Option {
+	return waitOption(wait)
+}
+
+func Fetch(opts ...Option) error {
+	options := &options{
+		baseURL:     baseURL,
+		dir:         filepath.Join(util.CacheDir(), "fetch", "openeuler", "csaf"),
+		retry:       5,
+		concurrency: 20,
+		wait:        1,
+	}
+	for _, o := range opts {
+		o.apply(options)
+	}
+
+	if err := util.RemoveAll(options.dir); err != nil {
+		return errors.Wrapf(err, "remove %s", options.dir)
+	}
+
+	client := utilhttp.NewClient(utilhttp.WithClientRetryMax(options.retry))
+	for _, kind := range []string{"cve", "advisories"} {
+		log.Printf("[INFO] Fetch openEuler %s Security Advisories (CSAF)", kind)
+		is, err := options.fetchCSAFIndex(client, kind)
+		if err != nil {
+			return errors.Wrapf(err, "fetch %s index", kind)
+		}
+		if err := options.fetchCSAF(client, kind, is); err != nil {
+			return errors.Wrapf(err, "fetch %s", kind)
+		}
+	}
+
+	return nil
+}
+
+func (o options) fetchCSAFIndex(client *utilhttp.Client, kind string) ([]string, error) {
+	u, err := url.JoinPath(o.baseURL, kind, "index.txt")
+	if err != nil {
+		return nil, errors.Wrap(err, "url join")
+	}
+
+	resp, err := client.Get(u)
+	if err != nil {
+		return nil, errors.Wrapf(err, "fetch %s", u)
+	}
+	defer resp.Body.Close() //nolint:errcheck
+
+	if resp.StatusCode != http.StatusOK {
+		_, _ = io.Copy(io.Discard, resp.Body)
+		return nil, errors.Errorf("error response with status code %d", resp.StatusCode)
+	}
+
+	var ls []string
+	scanner := bufio.NewScanner(resp.Body)
+	for scanner.Scan() {
+		ls = append(ls, scanner.Text())
+	}
+	if err := scanner.Err(); err != nil {
+		return nil, errors.Wrap(err, "scanner encounter error")
+	}
+
+	return ls, nil
+}
+
+func (o options) fetchCSAF(client *utilhttp.Client, kind string, is []string) error {
+	us := make([]string, 0, len(is))
+	for _, i := range is {
+		u, err := url.JoinPath(o.baseURL, kind, i)
+		if err != nil {
+			return errors.Wrap(err, "url join")
+		}
+		us = append(us, u)
+	}
+
+	if err := client.PipelineGet(us, o.concurrency, o.wait, func(resp *http.Response) error {
+		defer resp.Body.Close() //nolint:errcheck
+
+		switch resp.StatusCode {
+		case http.StatusOK:
+			if resp.ContentLength == 0 {
+				// e.g.
+				// https://repo.openeuler.org/security/data/csaf/cve/2019/csaf-openeuler-cve-2019-25219.json
+				// https://repo.openeuler.org/security/data/csaf/cve/2023/csaf-openEuler-cve-2023-52920.json
+				return nil
+			}
+
+			var csaf CSAF
+			if err := json.NewDecoder(resp.Body).Decode(&csaf); err != nil {
+				var syntaxErr *json.SyntaxError
+				if errors.As(err, &syntaxErr) {
+					// e.g.
+					// https://repo.openeuler.org/security/data/csaf/cve/2024/csaf-openeuler-cve-2024-50187.json
+					return nil
+				}
+				return errors.Wrap(err, "decode json")
+			}
+
+			switch kind {
+			case "cve":
+				splitted, err := util.Split(csaf.Document.Tracking.ID, "-", "-")
+				if err != nil {
+					fmt.Println(csaf.Document.Tracking.ID, err)
```

Comment:
for debug?

Reply:
fixed: https://github.com/MaineK00n/vuls-data-update/compare/029620ffa18965d9cd00104060d8e3fd029d4021..045a82642cd0cf480b2943cebe3ed369dfd09a9b